build: SDK and stampfile fixes / add tools tarball target

mcprat opened the pull request:
mostly taken from #9975 

fixes for the build system regarding dependencies and stampfiles and AUTOREMOVE

fixes for the SDK and rename of those tarball targets

proposal for a tarball for a prebuilt set of build tools

mcprat committed:
timestamp.pl: add options and cleanup syntax

A rewrite to improve the abilities of the script.

*Options*

 -x:	  exclude path pattern (find ... ! -path <>)
 -a:	  append option (add find operands)
 -colder: list only results older than <>
 -cequal: list only results with same mtime as <>
 -cnewer: list only results newer than <>
 -p:	  print only filenames
 -t:	  print only mtime
 -P:	  print basename as filename
 -T:	  print local time as mtime

*Usage*
(options above must come first)

 (path)		prints path and mtime in epoch time
 -n (path)	returns 0 if file exists, 1 if not
 -r (path)	recursive (if given path is directory)
 -- (list)	list of multiple files, also recursive
  -		read path from STDIN

*Examples*

scripts/timestamp.pl -n Makefile && echo "Makefile exists"
scripts/timestamp.pl -r scripts
scripts/timestamp.pl -- Makefile rules.mk
scripts/timestamp.pl -colder Makefile -- rules.mk
scripts/timestamp.pl -cnewer Makefile -- rules.mk
echo Makefile | scripts/timestamp.pl -

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: use and improve usage of timestamp.pl

Replace some non-POSIX usage of find with timestamp.pl
which is a Perl wrapper for find
where the output can be manipulated with options
(with or without it's mtime timestamp).

Using timestamp.pl with the '-n' option
was never and is still not able to process
more than one file at a time,
so use Make foreach function and split up lines
with logical operators.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
toolchain: cleanup dependencies

Reorganize dependencies between toolchain targets to save a line.
This also puts them in a linear order which is easier to read.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: define and use stampfile for CONFIG_AUTOREMOVE

In package.mk this stampfile is already used
but without a variable defined for it.

For consistency, define Make variables for
similar stampfiles in both host-build and package
to indicate that the autoremove rules took place.

Because the removal of files in the build directory
updates the timestamp of the directory itself,
and the build directory is used as a prerequisite
of one of the stampfiles by calling "rdep",
let this new stampfile be the target
when CONFIG_AUTOREMOVE is selected.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
target/tools: add target for prebuilt tools tarball

An automatic packaging of the staging_dir/host tools
including their stampfiles.

The resulting tar.xz can be unpacked in a new buildroot
in the top level directory (openwrt folder).

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: make autoremove rule a non-fatal error

On some drives mounted with buggy NFS,
a large number of deletion requests
causes a few random files with very long paths
to fail to be removed.

This feature is just for saving some space,
so failure to remove the files is not critical.

Make this command a non-fatal error
with the Make prefix '-'.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
toolchain: fix stampfile dependency

Use the gcc/final/install target stampfile as the prerequisite
for the toolchain/compile target stampfile.

The previous filename was also incorrect,
the stampfile name format was changed (see reference)
but was never updated here.

The issue was never noticed
because the usage of timestamp.pl was broken
(fixed in previous commit)
causing the extra prereq files passed to 'call stampfile'
to never get checked at all.

Fixes: 5149ed151 ("gcc: split up the build process into three distinct stages")
Signed-off-by: Michael Pratt <mcpratt@pm.me>

aparcar commented on target/tools/files/README.TOOLS:
```diff
@@ -0,0 +1,10 @@
+This should be unpacked directly in the top level directory (openwrt folder)
+example:
+
+	$ xzcat openwrt-tools.Linux-x86_64.tar.xz | tar xf - -C openwrt
```
Did you test that this actually recycles stamps? Meaning that you unpack it at varying path and it doesn't rebuild everything from tools/?

mpratt14 replied:
that is the plan, it needs more testing though

mpratt14 replied:
I tested with one package, xz so far, which does not rebuild

like this

```
make dirclean
make tools/xz/compile
make target/tools/compile NO_DEPS=1
cp bin/targets/ath79/generic/openwrt-tools.Linux-armv7l.tar.xz .
make dirclean
xzcat openwrt-tools.Linux-armv7l.tar.xz | tar xf -
make tools/xz/compile
```

Ansuel replied:
Wonder if I should test this on my github action 

Can you check if `make tools/install V=s` cause tools recompilation?

Ansuel replied:
@aparcar btw another option if we decide to host such tar in openwrt downloads would be use github actions cache 

But IMHO the idea of adding this to openwrt downloads will be NACK as we already have the sdk

Ansuel replied:
Mhhh don't know testing this and tools are recompiled.

```
find build_dir/host | xargs touch
sleep 1
find staging_dir/host/stamp | xargs touch
```

using this workaround tools are not recompiled.

Ansuel commented on include/package.mk:
```diff
@@ -258,7 +259,7 @@ define Build/CoreTargets
   ifneq ($(CONFIG_AUTOREMOVE),)
     compile:
-		-touch -r $(PKG_BUILD_DIR)/.built $(PKG_BUILD_DIR)/.autoremove 2>/dev/null >/dev/null
+		touch -r $(STAMP_BUILT) $(STAMP_REMOVED) 2>/dev/null >/dev/null
```
If you blame 6-7 years ago also this was added to the non fatal function. Wonder if we should keep this. I guess if STAMP_BUILD doesn't exist this will fail.

mpratt14 replied:
I cannot imagine it failing in good circumstances, since it would now use the variable for the stampfiles.

the "built" stampfile must exist before this is ran. An error would suggest that someone who is editing the makefile made some mistake where autoremove happens before the built stampfile is made.

maybe the `2>/dev/null` can be removed

mpratt14 replied:
I ended up making STAMP_REMOVED not rely on STAMP_BUILT at all

removal causes timestamp of the build directory to change, it would be useful to have a stampfile both before and after the timestamp of the directory

Ansuel commented on include/host-build.mk:
```diff
@@ -24,6 +24,7 @@ BUILD_TYPES += host
 HOST_STAMP_PREPARED=$(HOST_BUILD_DIR)/.prepared$(if $(HOST_QUILT)$(DUMP),,$(shell $(call find_md5,${CURDIR} $(PKG_FILE_DEPENDS),))_$(call confvar,CONFIG_AUTOREMOVE $(HOST_PREPARED_DEPENDS)))
 HOST_STAMP_CONFIGURED:=$(HOST_BUILD_DIR)/.configured
 HOST_STAMP_BUILT:=$(HOST_BUILD_DIR)/.built
+HOST_STAMP_REMOVED:=$(HOST_BUILD_DIR)/.removed
```
For package.mk this was added to call clean-build. For host-build.mk what is the use of such file? Consistency?

mpratt14 replied:
consistency, and who knows maybe actually used for something in the future

Ansuel replied:
also IMHO we should keep the autoremove naming just it's a nitpick

Ansuel commented:
@mpratt14 Can you put this in another pr? https://github.com/openwrt/openwrt/pull/10659/commits/e93903b351d642129e370bde9cb008a7b3879493
So i can merge this right away?

mpratt14 commented:
> Can you put this in another pr? [e93903b](https://github.com/openwrt/openwrt/commit/e93903b351d642129e370bde9cb008a7b3879493) So i can merge this right away?

I would but I don't think it's urgent...seems like that line is not used, otherwise it would have been discovered faster

Ansuel commented:
@mpratt14 in out of all my test i never experienced that part but considering the bug is clear and I can quickly merge it, it's a good idea and extra bonus point, we have less commit here :D

mpratt14 commented:
does it really need a new PR, or can you just grab it here

otherwise I'll do it later I have to go right now

https://github.com/mpratt14/openwrt/commit/e93903b351d642129e370bde9cb008a7b3879493.patch

mpratt14 commented:
ah wait.... it needs rebasing of course....

Ansuel commented:
@mpratt14 yep conflicts but np take your time :D

mpratt14 commented:
@Ansuel 
https://github.com/mpratt14/openwrt/commit/6e4e88627199a5a893e5146080b18b3d1aee910f

mpratt14 commented:
@Ansuel 

just did test with all tools, same buildroot

some random tools did recompile
first few lines of make log is like this

```
 make[1] tools/compile
 make[2] -C tools/flock compile
 make[2] -C tools/xz compile
 make[2] -C tools/sed compile
 make[2] -C tools/patch compile
 make[2] -C tools/tar compile
 make[2] -C tools/ninja compile
 make[2] -C tools/meson compile
 make[2] -C tools/pkgconf clean-build
 make[2] -C tools/pkgconf compile
```

note the `clean-build` which is unique to CONFIG_AUTOREMOVE (unless you command make to do that target)
which comes from `rebuild_check` in subdir.mk.
I think `rebuild_check` is deprecated and can be removed, but I'm not sure, because I have a hard time telling the difference between `.compile` target and `compile` target and `check-depends`, and why these distinctions are needed *only* when CONFIG_AUTOREMOVE is selected.

I can also verify that the hash in the stampfile's name is the same after rebuild as it is in the tarball

other than that, I timed `make tools/compile` when using the prebuilt tools package
something that on my system takes more than an hour with 1 job

```
real    8m31.136s
user    4m38.037s
sys     1m37.389s
```

and again my procedure to test was

```
make dirclean
make tools/compile
make target/tools/compile
cp bin/targets/*/*/openwrt-tools.Linux-aarch64.tar.xz .
make dirclean
xzcat openwrt-tools.Linux-aarch64.tar.xz | tar xf -
time make tools/compile
```

can you verify similar behavior when unpacking on the same buildroot? then we can worry about unpacking it on a different one.

mpratt14 commented:
so trying with a new buildroot for the first time I see the problem...

git doesnt care about timestamps, and every file in tools/ is timestamped for the moment it was made by git clone
the hash in the name of the "prepared" stampfile is made with those timestamps, so it's different

to make this work we have to completely replace the (first) hash with something else

I have ideas, but it might take a while...

xabolcs commented:
> git doesnt care about timestamps

Hi @mpratt14,
did you try `git-retore-mtime`?

- https://stackoverflow.com/a/13284558
- https://github.com/MestreLion/git-tools/

mpratt14 commented:
I'm not sure whether we should be using third-party tools to handle something like this, but we can certainly emulate what that tools does with a script or something

the real problem is the hash, a user might change a Makefile and then put it back, and then build the tools tarball and try to use it on another buildroot which has a different (but even older) timestamp. we should be relying on a difference in timestamps instead of a hash of them

mcprat committed:
build: replace stampfile timestamp hash with comparison

Using a hash to represent the timestamps of dependencies
for each submake build is gratuitous and unnecessary.
It also blocks the ability to work with the build system
across different buildroots or across different systems.

A rebuild only needs to take place
if the dependencies have updated
with respect to the stampfile that represents them
based on the difference between the timestamps,
just like the way that Make behaves.

For host-build and packages, replace the hash
with a new requirement for the dependency check
that compares the latest timestamp of the depending files
with the timestamp of the stampfile that is associated with them.

The timestamps of dependencies must be less than or equal to
the timestamp of the stampfile. Otherwise,
<stampfile>_check will be touched,
triggering the rules of that target later on.

For now, the kernel still uses a hash of timestamps,
but the kernel is a product of the build system,
expected to be rebuilt more often and not be recycled.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
rebased

@Ansuel I think i figured out how to remove the hash safely, with the smallest diff, originally I thought I would have to change a lot of things

this should work now across buildroots, but using a small script that I added to touch everything that is unpacked
instead of trying to get git to restore timestamps, we just make the stampfiles "newer"

I only tested xz on the same buildroot so far

xabolcs commented:
> ... using a small script that I added to touch everything that is unpacked ...

GNU `tar --touch` does exactly this and GNU `tar` is a prerequisite.
```bash
# tar --help|grep touch
  -m, --touch                don't extract file modified time
```

mpratt14 commented:
ah thats nice, I didnt look hard enough for that 😆

xabolcs commented:
> ah thats nice, I didnt look hard enough for that laughing

It weren't there until you wrote your touch tool! :rofl:

Ansuel commented:
@mpratt14 fun thing -m is not usable :D 
I tried A LOT... many inconsistencies... tar tens to extract staging_dir first and then build_dir so tools gets recompiled... also touching build_dir files is still needed for some reason or again... tools are recompiled...

xabolcs commented:
Those ordering flags are unusable too?

```
  -s, --preserve-order, --same-order
                             member arguments are listed in the same order as
                             the files in the archive
      --sort=ORDER           directory sorting order: none (default), name or
                             inode
```

xabolcs commented:
@mpratt14 
```
... xargs -I {} touch "$dir/{}" ...
```

How long is your script is running?
Those reference options of `touch` wouldn't be needed?
```
  -r, --reference=FILE   use this file's times instead of current time
  -t STAMP               use [[CC]YY]MMDDhhmm[.ss] instead of current time
```

mpratt14 commented:
> tar tends to extract staging_dir first and then build_dir

that should not make a difference
however, the build directories have to be made / touched, before the stampfiles within it

-m is part of unix specification

should be able to do `tar -xmf` ??

script is pretty fast, but if tar works like described we wont need it

Ansuel commented:
> Those ordering flags are unusable too?
> 
> ```
>   -s, --preserve-order, --same-order
>                              member arguments are listed in the same order as
>                              the files in the archive
>       --sort=ORDER           directory sorting order: none (default), name or
>                              inode
> ```

tested that... no change :(

Ansuel commented:
> that should not make a difference

thing is that if we need to touch files in build_dir (not a lot we use AUTOREMOVE) and bin in staging_dir then we can skip -m option as the timestamp will be modified anyway.

The correct solution would be find the correct way to NOT touch the file at all

mpratt14 commented:
the problem is that in a new repo (git clone) all the files are touched
the stampfiles in build_dir must be newer than the Makefiles, otherwise a rebuild

also, you need to make a new tarball, since the prepared stampfile hash is gone now

mpratt14 commented:
so now I'm experimenting with touching the stampfiles in order, which seems to finally work

mcprat committed:
build: define make variable HOST_SKIP_DOWNLOAD

Define the variable HOST_SKIP_DOWNLOAD
in order to have more control over calling download targets
which set the download file of a package
as a prerequisite to prepare target stampfiles.

Use the conditions already defined in host-build.mk
as the default definition.

Targets for downloading can then be skipped manually
by defining HOST_SKIP_DOWNLOAD in an individual package's Makefile
or by defining it at the command line or shell environment.

This is consistent with the existing variable in package.mk
PKG_SKIP_DOWNLOAD, with the same functionality,
but different default values.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
OK, so now, we have it working on a new buildroot, with exactly expected behavior, and I didn't have to remove `rebuild_check`

the final problem was, `$(call Download,default)` causes the download file to be a prerequisite of `.prepared` stampfiles,
so even with that stampfile being there, it's prerequisite was missing, triggering a rebuild

This causes us to need a way to skip that manually, so we now have `HOST_SKIP_DOWNLOAD` flag

Unfortunately, the stampfiles still have to be touched in order with a script, otherwise we would need a way to ignore the stampfile targets completely, but I'm aiming for the smallest diff possible because of how sensitive the build system can be here. However at the same time, it should check *something* and not just fully trust that all tools are present. 

Also, renamed that script

I test on a new repo like this (after first using this branch to build the tools tarball)

```
git clone -b stampfile-timestamp-tools https://github.com/mpratt14/openwrt.git
cd openwrt
make defconfig; make dirclean; make prereq
scripts/unpack_stamp.sh ../openwrt-tools.Linux-aarch64.tar.xz .
time make tools/compile CONFIG_AUTOREMOVE=y HOST_SKIP_DOWNLOAD=1
```

mpratt14 commented:
alternative to touching the stampfiles in order would be to use `CONFIG_AUTOREBUILD` to set "order-only" prerequisites when disabled. but I don't know if that would work yet

https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html

mpratt14 commented:
I forgot to rename the script in other commits anyway...

I'll try to get rid of the script, the "order-only" rules work, so it's good (as long as older Make versions support it...)

mpratt14 commented:
@Ansuel ready for review

my test procedure:

making the tarball with this PR

```
rm .config
make defconfig; make dirclean; make prereq
time make tools/compile target/tools/compile CONFIG_AUTOREMOVE=y
```

using it on a new buildroot (also this PR)

```
git clone -b stampfile-timestamp-tools https://github.com/mpratt14/openwrt.git
cd openwrt
make defconfig; make dirclean; make prereq
make menuconfig # or nano .config (disable CONFIG_AUTOREBUILD)
cd ..
xzcat openwrt-tools.Linux-aarch64.tar.xz | tar -xf - -C openwrt
cd openwrt
time make tools/compile CONFIG_AUTOREMOVE=y HOST_SKIP_DOWNLOAD=1
```

and note that the tools stampfile is now touched after tools/compile or tools/install command

```
find staging_dir | grep tools_compile
```

next thing to do:

 - add tools stampfile as prereq to target/tools so we can do

```
make defconfig; make dirclean; make prereq
time make target/tools/compile CONFIG_AUTOREMOVE=y
```

and target/tools/compile will cause tools/compile in a submake

mpratt14 commented:
tools/compile stampfiles now get checked in less than a minute

```
time make tools/compile CONFIG_AUTOREMOVE=y HOST_SKIP_DOWNLOAD=1

...
...

real    0m36.950s
user    0m19.423s
sys     0m6.652s
```

mcprat committed:
build: make find_md5 reproducible and hash content

Currently, find_md5 is a hash of timestamps.
Now that there is a separate way to handle timestamps,
finding the newest timestamp with find_ts
and comparing that with the stampfile's timestamp,
there is no longer a need to record a list of timestamps
in the form of a hash of them.

Also, hashing timestamps is an inferior method for stampfiles
because it is possible for a timestamp to update
(or even be forced to be set to an old time)
without any file content changing.

Git does not preserve timestamps in any way,
therefore, by having find_md5 be a hash
of the content of the list of files given,
it is reproducible across different buildroots and systems
using the same versions of each file in the list.

A hash of the hashes still represents the same state of the file list
but in a short consistent form not dependent on the number of files.
The hashes are sorted so that the final hash
is not dependent on the alphabetical order of the input files.

Suggested-by: Christian Marangi <ansuelsmth@gmail.com>
Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
build: reintroduce prepare stage hash

Now that find_md5 is a hash of file contents
and not a hash of timestamps,
a hash for keeping track of changes in depending files
can be reintroduced, based on CONFIG_AUTOREBUILD,
for rebuilding whenever the hash changes.

Instead of including this in the definition
of the prepare stampfile, or having multiple defines,
have a variable for the hash itself,
and append it to the name of the stampfile
based on CONFIG_AUTOREBUILD,
for each dependency and as a duplicate target.
This allows for changing CONFIG_AUTOREBUILD
without causing everything to rebuild
just because of the change in configuration.

Let the function rdep check for timestamps first
and compare the hashes if that fails,
allowing for the stampfile hash to prevent a full rebuild
if timestamps of depending files change
but not their content.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
I removed the commits related to the SDK (to another branch for later)

I discovered #10421 and I implemented the first commit there in a different way in the last 2 commits here

first: to make find_md5 a hash of the file content, exactly the same way
second: to implement the hash back into `host-build.mk` and `package.mk`, but based on `CONFIG_AUTOREBUILD`
and in a completely different way so the config can change

last 2 commits need testing

mpratt14 commented:
have to figure out what is wrong later

this line in `toplevel.mk` is troublesome

```
$(eval $(call rdep,scripts/config,scripts/config/mconf))
```

mcprat committed:
build: split download reverse dependencies

Define DOWNLOAD_RDEP based on which build type is being done.
It could be both host-build and package build, so concatenate.

This allows the value of DOWNLOAD_RDEP to be controlled
with each build-type, consistent with how
targets and prerequisites are set in different makefiles
according to the build type for stampfiles.

Let the download file be a prerequisite of a stampfile
only when downloading is not being skipped.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mcprat committed:
config: move rdep call to config Makefile

This line can no longer execute correctly in toplevel.mk
it must be executed after MKHASH is built and defined.

Adding rules.mk has the extra benefit
of silencing some output and printing submake info.

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
ok, this PR is in a working state again, and ready for review...again

this time, it works with or without CONFIG_AUTOREBUILD
commit messages and README still needs editing

current test procedure:

creating tarball:
```
rm .config
make defconfig; make dirclean; make prereq
time make target/tools/compile CONFIG_AUTOREMOVE=y
```

using it in another buildroot
```
git clone -b stampfile-timestamp-tools https://github.com/mpratt14/openwrt.git
cd openwrt
make defconfig; make dirclean; make prereq
cd ..
xzcat openwrt-tools.Linux-aarch64.tar.xz | tar -xf - -C openwrt
time make tools/compile CONFIG_AUTOREMOVE=y HOST_SKIP_DOWNLOAD=1
```

its faster with CONFIG_AUTOREBUILD disabled
on my system, enabled vs disabled:

```
real    2m43.663s
user    1m40.850s
sys     0m39.351s

real    1m8.536s
user    0m38.301s
sys     0m15.701s
```

mpratt14 commented:
splitting a commit to another PR, a better way to handle `.tools_compile_...` getting touched

cleaned up new config help and new readme

dhewg commented:
I hope this isn't unrelated, but it may be related to timestamps too:
If I build some target, then wipe build_dir, all of the tools are getting rebuilt. But they're all still there in staging_dir.
Is that solvable or some required dependency that isn't obvious?

mpratt14 commented:
openwrt build system uses "stampfiles" as the build target for each tool and package (literally the "target" of each Makefile, in Make terminology) which is just an empty file that gets touched after every other command for that stage of building is successful

the stampfiles are stored in build_dir, and if they are deleted, Make considers those things to not be built at all on the next run, so they get rebuilt

there is a stampfile in staging_dir for each package, but that is the end of the dependency chain, not the beginning, in other words, installing into staging_dir depends on building in build_dir, not the other way around. When a dependency is missing, *everything* in that dependency chain is considered out-of-date.

if you want to delete build_dir after builds you have to keep all the empty files

`find build_dir -type f '!' -size 0 | xargs rm -rf`

or you use CONFIG_AUTOREMOVE which does this for you in each package build directory after build is done

dhewg commented:
That's clear so far, but if the build process requires e.g. `flock`, a dependency on `./staging_dir/host/stamp/.flock_installed` should be enough to ensure that it's installed (which implies an earlier build step), right?

I guess that there's some dep on `./build_dir/host/flock-2.18/.built` too, but that should only be a dependency when installing it, not to make sure it is already installed, right? And if that were given, it would solve my 'problem'

mpratt14 commented:
Make looks for what needs to be updated whenever it is ran. `./staging_dir/host/stamp/.flock_installed` does imply that at some point the dependencies were handled, but Make doesn't work on those implications, only on the existence and timestamp of the target and its dependencies, and that is critical to the design of Make. However, you could argue against the use of so many stampfiles here and why they are in build_dir instead of staging_dir... and I don't know exactly why that is but it seems to be for ease of development rather than ease of use. At the very least, when something errors you can see exactly which build stage the error occured in without scrolling through the logs.

the whole idea of a "dependency" in Make terms is that, when B depends A, A must be built and up-to-date, otherwise we cannot consider B to even be built at all. The way we would naturally think of "dependency" is more like "A was required in order to achieve B, and now we have B so there's nothing to do with A". There's no other way for Make to know whether rebuilding A makes a difference in B, other than the fact that A changed or is missing. (in this PR the hash on the stampfile is now a fallback for the timestamps, but this hack in the "rdep" function still uses stampfiles and touches them to update a timestamp)

mpratt14 commented:
> that should only be a dependency when installing it, not to make sure it is already installed

in Make terminology, these 2 ideas are the same

dhewg commented:
> and why they are in build_dir instead of staging_dir

Of course I started looking at it just after commenting here, and I noticed exactly that:
```
diff --git a/include/host-build.mk b/include/host-build.mk
index f06dd80230..3685ddf42a 100644
--- a/include/host-build.mk
+++ b/include/host-build.mk
@@ -21,10 +21,10 @@ include $(INCLUDE_DIR)/depends.mk
 include $(INCLUDE_DIR)/quilt.mk
 
 BUILD_TYPES += host
-HOST_STAMP_PREPARED=$(HOST_BUILD_DIR)/.prepared$(if $(HOST_QUILT)$(DUMP),,$(shell $(call find_md5,${CURDIR} $(PKG_FILE_DEPENDS),))_$(call confvar,CONFIG_AUTOREMOVE $(HOST_PREPARED_DEPENDS)))
-HOST_STAMP_CONFIGURED:=$(HOST_BUILD_DIR)/.configured
-HOST_STAMP_BUILT:=$(HOST_BUILD_DIR)/.built
 HOST_BUILD_PREFIX?=$(if $(IS_PACKAGE_BUILD),$(STAGING_DIR_HOSTPKG),$(STAGING_DIR_HOST))
+HOST_STAMP_PREPARED=$(HOST_BUILD_PREFIX)/stamp/.$(PKG_NAME)_prepared$(if $(HOST_QUILT)$(DUMP),,$(shell $(call find_md5,${CURDIR} $(PKG_FILE_DEPENDS),))_$(call confvar,CONFIG_AUTOREMOVE $(HOST_PREPARED_DEPENDS)))
+HOST_STAMP_CONFIGURED:=$(HOST_BUILD_PREFIX)/stamp/.$(PKG_NAME)_configured
+HOST_STAMP_BUILT:=$(HOST_BUILD_PREFIX)/stamp/.$(PKG_NAME)_built
 HOST_STAMP_INSTALLED:=$(HOST_BUILD_PREFIX)/stamp/.$(PKG_NAME)_installed
 
 override MAKEFLAGS=
@@ -143,7 +143,7 @@ ifndef DUMP
   $(if $(DUMP),,$(call HostHost/Autoclean))
 
   $(HOST_STAMP_PREPARED):
-   @-rm -rf $(HOST_BUILD_DIR)
+   @-rm -rf $(HOST_BUILD_DIR) $(HOST_STAMP_BUILT) $(HOST_STAMP_CONFIGURED) $(HOST_STAMP_PREPARED)
    @mkdir -p $(HOST_BUILD_DIR)
    $(foreach hook,$(Hooks/HostPrepare/Pre),$(call $(hook))$(sep))
    $(call Host/Prepare)
@@ -190,7 +190,7 @@ ifndef DUMP
 
   host-clean-build: FORCE
    $(call Host/Uninstall)
-   rm -rf $(HOST_BUILD_DIR) $(HOST_STAMP_BUILT)
+   rm -rf $(HOST_BUILD_DIR) $(HOST_STAMP_BUILT) $(HOST_STAMP_CONFIGURED) $(HOST_STAMP_PREPARED)
 
   host-clean: host-clean-build
    $(call Host/Clean)
@@ -198,8 +198,7 @@ ifndef DUMP
 
     ifneq ($(CONFIG_AUTOREMOVE),)
       host-compile:
-       $(FIND) $(HOST_BUILD_DIR) -mindepth 1 -maxdepth 1 -not '(' -type f -and -name '.*' -and -size 0 ')' -print0 | \
-           $(XARGS) -0 rm -rf
+       rm -rf $(HOST_BUILD_DIR) $(HOST_STAMP_BUILT) $(HOST_STAMP_CONFIGURED) $(HOST_STAMP_PREPARED)
     endif
   endef
 endif
```

Only very lighty tested, but it *seems* to be doing what I want it to do :)

It decouples built tools/ from build_dir, which may even be required for that binary-tools-download-package?

And it's a bit tidier too:
```
$ find .|grep  stamp|grep flock
./staging_dir/host/stamp/.flock_installed
./staging_dir/host/stamp/.flock_built
./staging_dir/host/stamp/.flock_configured
./staging_dir/host/stamp/.flock_prepared003df113fc9390817fb2caff4bc1d090_6664517399ebbbc92a37c5bb081b5c53
```

mpratt14 commented:
yeah that edit looks good, but I'm sure one of the members would point out that there's some reason we don't want to do that. just not sure what that reason is and I don't want to bother with it on top of everything else I'm trying to do 😆

dhewg commented:
Aye, I'll poke it some more and if it works out I'll throw it in a PR. Thanks!

mcprat committed:
reproducible remove fixup

Signed-off-by: Michael Pratt <mcpratt@pm.me>

mpratt14 commented:
I'm going to split this PR because another PR depends on the fixes here before the last commit
